converter/codec.go

```diff
@@ -342,27 +342,91 @@ func NewPayloadCodecHTTPHandler(e ...PayloadCodec) http.Handler {
 	return &codecHTTPHandler{codecs: e}
 }
 
-// RemoteDataConverterOptions are options for NewRemoteDataConverter.
+// RemotePayloadCodecOptions are options for RemotePayloadCodec.
 // Client is optional.
-type RemoteDataConverterOptions struct {
+type RemotePayloadCodecOptions struct {
 	Endpoint      string
 	ModifyRequest func(*http.Request) error
 	Client        http.Client
 }
 
-// remoteDataConverter is a DataConverter that wraps an underlying data
-// converter and uses a remote codec to handle encoding/decoding.
+type remotePayloadCodec struct {
+	options RemotePayloadCodecOptions
+}
+
+func NewRemotePayloadCodec(options RemotePayloadCodecOptions) PayloadCodec {
+	return &remotePayloadCodec{options}
+}
+
+func (pc *remotePayloadCodec) Encode(payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
+	return pc.encodeOrDecode(pc.options.Endpoint+remotePayloadCodecEncodePath, payloads)
+}
+
+func (pc *remotePayloadCodec) Decode(payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
+	return pc.encodeOrDecode(pc.options.Endpoint+remotePayloadCodecDecodePath, payloads)
+}
+
+func (pc *remotePayloadCodec) encodeOrDecode(endpoint string, payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
+	requestPayloads, err := json.Marshal(commonpb.Payloads{Payloads: payloads})
+	if err != nil {
+		return payloads, fmt.Errorf("unable to marshal payloads: %w", err)
+	}
+
+	req, err := http.NewRequest(http.MethodPost, endpoint, bytes.NewReader(requestPayloads))
+	if err != nil {
+		return payloads, fmt.Errorf("unable to build request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	if pc.options.ModifyRequest != nil {
+		err = pc.options.ModifyRequest(req)
+		if err != nil {
+			return payloads, err
+		}
+	}
+
+	response, err := pc.options.Client.Do(req)
+	if err != nil {
+		return payloads, err
+	}
+	defer func() { _ = response.Body.Close() }()
+
+	if response.StatusCode == 200 {
+		var resultPayloads commonpb.Payloads
+		err = jsonpb.Unmarshal(response.Body, &resultPayloads)
+		if err != nil {
+			return payloads, fmt.Errorf("unable to unmarshal payloads: %w", err)
+		}
+		if len(payloads) != len(resultPayloads.Payloads) {
+			return payloads, fmt.Errorf("received %d payloads from remote codec, expected %d", len(resultPayloads.Payloads), len(payloads))
+		}
+		return resultPayloads.Payloads, nil
+	}
+
+	message, _ := io.ReadAll(response.Body)
+	return payloads, fmt.Errorf("%s: %s", http.StatusText(response.StatusCode), message)
+}
+
+// RemoteDataConverterOptions are options for NewRemoteDataConverter.
+type RemoteDataConverterOptions struct {
+	RemotePayloadCodecOptions
```

Comment:
This is a backwards incompatible change. I know it may seem strange (especially w/ code reuse in other languages), but I think the fields need to be copied in here same as they are in `RemotePayloadCodecOptions`. In addition to this being a backwards incompatible change, a user shouldn't have to know that we are reusing remote codec code underneath.

Reply:
Good catch @cretz! Still getting used to Go's embedded fields etc. Fixed: the fields are now present in both structs.